dateutil/test/test_parser.py

```diff
@@ -10,6 +11,35 @@
 from six import assertRaisesRegex, PY3
 from six.moves import StringIO
 
+import pytest
+
+
+# If we have a 4-digit year, a non-numeric month (abbreviated or not),
+# and a day (1 or 2 digits), then there is no ambiguity as to which
+# token is a year/month/day.  This holds regardless of what order the
+# terms are in and for each of the separators below.
+seps = ['-', ' ', '/', '.']
+token_opts = [['%Y'], ['%b', '%B'], ['%d', '%-d']]
+
+prods = itertools.product(*token_opts)
+perms = [y for x in prods for y in itertools.permutations(x)]
```

Comment:
Not sure if this is more readable per se, but I believe this is equivalent to:

`list(itertools.chain.from_iterables(map(itertools.permutations, itertools.prod(*token_opts))))`

or possibly better:

`list(itertools.chain(*(itertools.permutations(x) for x in itertools.prod(*token_opts))))`